students/GuyCChan/session07/html_render.py

```diff
@@ -50,20 +75,79 @@ class Head(Element):
 class OneLineTag(Element):
     pass
 
-    def render(self, out_file, **attributes):
-        open_tag.write("<{} ".format(self.tag))
-        for key, value in self.attributes.items():
-            open_tag.write("{}='{}'".format(key, value))
-        open_tag.append(">\n")
+    def render(self, out_file, cur_ind="", **kwargs):
+        open_tag, close_tag = self._open_close_tags()
+        out_file.write(cur_ind)
+        out_file.write(open_tag)
         out_file.write(self.contents[0])
-        out_file.write("</{}>\n".format(self.tag))
+        out_file.write(close_tag)
+        out_file.write("\n")
 
     def append(self, content):
+        """ Cannot add content to OneLineTag."""
         raise NotImplementedError
 
 
 class Title(OneLineTag):
     tag = "title"
 
 
+class SelfClosingTag(Element):
+
+    def __init__(self, content=None, **kwargs):
+        if content is not None:
+            raise TypeError("SelfClosingTag cannot contain any content.")
+        super().__init__(content=content, **kwargs)
+
+
+    def render(self, out_file, cur_ind="", **kwargs):
+        # Please explain the use of the underscore in the following line. I tried and tried for hours
+        # and over 2 days to get this going but failed. Is it the so called "throwaway" variable?
+        # How does it work here? Without it, tests for "hr" would always come back saying "hr" has no
+        # attribute 'open_tag'.
+        open_tag, _ = self._open_close_tags()
+        # This line didn't work for me: tag = self._open_tag()[:-1] + " />\n"
+        out_file.write(cur_ind)
+        out_file.write(open_tag.replace(">", " />\n"))
+
+    def append(self, *args):
+        """ Cannot add content to a SelfClosingTag."""
+        raise TypeError("You cannot add content to a SelfClosingTag")
+
+class Hr(SelfClosingTag):
+    tag = "hr"
+
+
+class Br(SelfClosingTag):
+    tag = "br"
+
+
+class A(OneLineTag):
+    tag = "a"
+
+    def __init__(self, link, content=None, **kwargs):
+        kwargs['href'] = link
+        super().__init__(content, **kwargs)
+
+
+class Li(Element):
+    tag = "li"
+
+
+class Ul(Element):
+    tag = "ul"
+
+
+class H(OneLineTag):
+    tag = "h"
+
+    def __init__(self, level, *args, **kwargs):
+        # Please explain why the param "content" has to be replaced with *args
+        # in order to have it work. Why positional arguments?
```

Comment:
you could use "content" here instead, in both places:

```
    def __init__(self, level, content=None, **kwargs):
        self.tag = "h" + str(level)
        super().__init__(content, **kwargs)
```
but you need the `=None` as it's an optional argument.

but the *args, **kwargs approach is a way to be clear that you are passing on exactly what is passed in to the superclass' __init__, without changes. It makes it clear that the API is the same except for the one added "level" parameter.

It also makes it more robust in the face of changes in the superclass -- if we changed the __init__ in the superclass to require one, two (or zero) positional parameters, then *args would still work.